### DISCUSSION TITLE
Style Guide: Short-form Lists

### DISCUSSION THREAD

**Discussion by NotTheDr01ds** [nushell/nushell#14221] [Polls]
The [Best Practices](https://nushell.sh/book/style_guide.html) (a.k.a., Style Guide) chapter of the Nushell Book includes recommendations for Nushell code that is contributed to official Nushell repositories (e.g., [nu_scripts](/nushell/nu_scripts) as well as the documentation).

As mentioned in the first paragraph:

> as for now it is still work in progress and subject to change. We welcome discussion and contributions.

On this week's call, we agreed that we need to more actively reach out for community feedback on certain areas of style preference.  This is one of those areas.

Short lists in Nushell (those displayed on a single line) can have several forms:

1. Option 1 - No extra space inside brackets; list items are separated by single spaces only:

   ```nu
   [item1 item2 item3]
   ["Item 1" "Item 2" "Item 3"]
   ```

2. Option 2 - Single-space inside brackets; list items are separated by single spaces only:

   ```nu
   [ item1 item2 item3 ]
   [ "Item 1" "Item 2" "Item 3" ]
   ```

3. Option 3 - No extra space inside brackets; list items are separated by a comma, followed by a space:

   ```nu
   [item1, item2, item3]
   ["Item 1", "Item 2", "Item 3"]
   ```

4. Option 4 - Single space inside brackets; list items are separated by a comma, followed by a space:

   ```nu
   [ item1, item2, item3 ]
   [ "Item 1", "Item 2", "Item 3" ]
   ```

5. Option 5 - We don't need to enforce a common style for this. It is okay if some code uses one form, and other code in the same module/repo uses another.

---

Note:  This is only for "short-form" lists.  It's recommended that longer lists be multi-line.

**Comment by NotTheDr01ds** [comment]
Update - Poll was reset and one vote erased (apologies to the voter).  This poll now only focuses on lists.  Records are a separate discussion with too many permutations to cover in one poll.

**Comment by NotTheDr01ds** [comment]
One reason I tossed the "don't need to enforce" option in is because I think there are a lot of cases where "readability" is determined by the content in the list, and it should be left to the judgement of the author.

I typically prefer spaces between bareword strings, ints, and some other types, but quoted strings (IMHO) are more readable with commas:

For instance, I like:

```nu
[ item1 item2 item3 ]
# but
["Item 1", "Item 2", "Item 3"]
```

The alternatives to me just aren't as "pretty":

```nu
[item1 item2 item3]
# or
["Item 1" "Item 2" "Item 3"]
```

I feel like the addition of the quotes pretty much dictates the need for commas.

If we have to land on one or the other, I opt for commas, since they will be more useful in general.  But I'd almost rather leave it up to the judgement of the person writing the code in this case.

**Comment by lemorage** [comment]
As a personal taste or a custom or whatever, I could not live without commas, when putting items into a list. Option 3 or 4 is either fine to me, but I definitely couldn’t accept absolutely no style enforcement; it’s a disaster to read and certainly not a good convention for newcomers. If we only allow both 3 and 4 OR both 1 and 2, it might be a good tempt. Not all of them, however.

**Comment by ghost** [comment]
my personal go-to is to put leading and trailing spaces into code (lambdas, etc) (`{ git commit }`), but not data (`[foo]`, `{a: 1}`), which improves readability for me (yes this is more relevant for dicts, but if dicts have it lists should as well).  
in terms of commas i decide based on the data (`[foo, bar]`, `[1, 2]`, `[1 2 3 4 5 6]`, `[{a: 1} {a: 2}]`).

**Comment by NotTheDr01ds** [comment]
> in terms of commas i decide based on the data

Agreed - that's the point I was going for in "don't enforce" as well.  The more complex the data in a list, the likely it is to need to be comma-delimited, IMHO.

We could, alternatively, define this as a more explicit rule regarding the types and when to use commas, but I'm feeling we're already getting pretty "in the weeks" with some of these rules.

On second thought - It might not be *too* bad. I think the only data-in-a-list that should be space-separated (no commas) is:

* Bareword strings
* Ints
* Variables (`[$a $b $c]`)
* Booleans
* And maybe floats (but I lean to "no")

With all others needing commas.

Would folks be open to that?

**Reply by ghost** [reply]
while this might make sense its also a lot to remember when not having a formatter.  
yes its only 5 things, but it might expand (what about file-sizes, dates, etc?) and there's already a lot of other things to keep in mind.

**Reply by NotTheDr01ds** [reply]
> when not having a formatter.

I'm assuming that whatever we land on will become the default for `nufmt` eventually.  While we can always update the rules then, I always worry that the ["follow the horse's ass"](https://www.reddit.com/r/FreightBrokers/comments/1bntrtf/a_horses_arse_a_somewhat_true_story_about_rail/) rule will prevail ;-)

**Reply by NotTheDr01ds** [reply]
Side-note:  It shouldn't be too hard to remember either, with the simplified version of the rule being *"list of purely alpha-numeric tokens (or variable names) can use a space"* but *"lists with any non-alphanumeric character in the list should be comma-separated"*

E.g., `[one two 4 5]`, but `[1.554, 3.777, 8.999]`

And side-note, yes, that would pertain to file-sizes (int file-sizes are alpha-num, but floats are not) and dates (only nanoseconds-since-the-epoch ints would be allowed with spaces).
